Automattic/publish-to-s3-gradle-plugin
# Adds support for Android publishing block

The goal of this PR is to address the following warning from Android Gradle Plugin:

```
WARNING:Software Components will not be created automatically for Maven publishing from Android Gradle Plugin 8.0. To opt-in to the future behavior, set the Gradle property android.disableAutomaticComponentCreation=true in the `gradle.properties` file or use the new publishing DSL.
```

The way to address this within a specific project is to add the following DSL:

```
android {
    publishing {
        singleVariant('release') {
            withSourcesJar()
            withJavadocJar()
        }
    }
}
```

We'd need to individually add this for every module that we publish. Since we already have our own plugin to unify publishing of our repositories, I thought it'd be better to handle it here. Here is a list of all the changes:

* 50609f4be817ebd3c8062bc4ff82210ec7547055 updates Gradle to `7.4.2` which is the version we currently use in our projects - although we are in the process of updating them to `8.2.1`. I tried updating this project to `8.2.1`, but then there are some issues with the AGP dependency and I thought this was the more simpler solution for now.
* 437ccb9d47156f9ac3c23fc11537235c9c4f7085 adds `local.properties` to `.gitignore`
* df6adaccfa23caf1a352428feeefad18c1656092 updates Kotlin `jvm` plugin to `1.6.10`. It's important that we avoid updating this to a version that is newer than the version of Kotlin in any of the projects that we use this plugin to publish it with. Otherwise, we might get unexpected failures while publishing those projects. I _think_ `1.6.10` is a safe bet and if a project is even more behind than that we could consider updating the Kotlin version of that project first.
* 9de4ec25fd18b6b37fd2590a76002a876787984d updates AGP dependendency to `7.2.1`. This is also a good one to keep outdated as much as possible to increase our range of projects that we can support. I had to update it to in `7+` range though because the `publishing` block in AGP wouldn't have been available otherwise, so I picked `7.2.1` which is a version we used in our projects for a long time. Note that I did try updating it to `8.0.2` first, but there was a host of issues with its compatibility with Gradle & Kotlin version.
* 9ce53b6d4781e0a8d52e564cb2cf0efd1227448c adds the `publishing` block shared at the beginning of this PR. It also removes the `androidSourcesJar` task since this is now covered by the added `publishing` block.
* 285093458b743bc0d604023de89bac5dd031fe58 updates the version to `0.8.0-rc-5`. This version is temporary as I'll update it to `0.8.0` before we merge this PR. Unfortunately we still don't have CI integration in this repo yet, so I had to manually publish these versions and I got to `-rc-5` until I was happy with it 😅 

**To Test**

I prepared a couple test PR for [WordPress-MediaPicker-Android](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/pull/70) & [wellSql](https://github.com/wordpress-mobile/wellsql/pull/26) to verify the changes.

* Notice in [this build](https://buildkite.com/automattic/wordpress-mediapicker-android/builds/210#01892af4-08bb-4b69-b7fe-38f98f0ff169/169-191) from [`WordPress-MediaPicker-Android`'s `trunk` at this time](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/tree/323497d4d810a547f2bf5bb84ec8193a12bc57de) has the warning about software components whereas [this build](https://buildkite.com/automattic/wordpress-mediapicker-android/builds/219#01897555-1671-4726-b66c-48aa15a0883a/173-192) from the [test PR](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/pull/70) does not.
* Similarly [this build](https://buildkite.com/automattic/wellsql/builds/24#01892ab4-9f5f-423c-a48e-2de0b326b4b0/169-191) from [`wellsql`'s `trunk` at this time](https://github.com/wordpress-mobile/wellsql/tree/5cfd1623a25880c16dc16852812fbf77de5758ba) has the warning about software components whereas [this build](https://buildkite.com/automattic/wellsql/builds/30#018975b2-caa9-48b7-b997-3f3142dca15f/173-193) from the [test PR](https://github.com/wordpress-mobile/wellsql/pull/26) does not.
* In order to verify that the `-sources.jar` & `-javadoc.jar` were published, I checked our S3 Maven directly. Here are the [`-sources.jar`](https://a8c-libs.s3.amazonaws.com/android/org/wordpress/mediapicker/70-bff66e5c6bb70dbf218296624dc386fcd2e3ea25/mediapicker-70-bff66e5c6bb70dbf218296624dc386fcd2e3ea25-sources.jar) & [`-javadoc.jar`](https://a8c-libs.s3.amazonaws.com/android/org/wordpress/mediapicker/70-bff66e5c6bb70dbf218296624dc386fcd2e3ea25/mediapicker-70-bff66e5c6bb70dbf218296624dc386fcd2e3ea25-javadoc.jar) files for `WordPress-MediaPicker-Android` & and here are the [`-sources.jar`](https://a8c-libs.s3.amazonaws.com/android/org/wordpress/wellsql/26-df99a0d755ea4fc7cf5d6b3d7f4a2f4aeac13f35/wellsql-26-df99a0d755ea4fc7cf5d6b3d7f4a2f4aeac13f35-sources.jar) & [`-javadoc.jar`](https://a8c-libs.s3.amazonaws.com/android/org/wordpress/wellsql/26-df99a0d755ea4fc7cf5d6b3d7f4a2f4aeac13f35/wellsql-26-df99a0d755ea4fc7cf5d6b3d7f4a2f4aeac13f35-javadoc.jar) files for `wellsql`.
* I also verified that the source files were published for `WordPress-MediaPicker-Android`, by updating the `mediaPickerVersion` for WCAndroid to `70-bff66e5c6bb70dbf218296624dc386fcd2e3ea25` and then using Android Studio's `cmd + click` feature to open the source file of an `enum`: [`MediaPickerContext`](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/blob/trunk/mediapicker/domain/src/main/java/org/wordpress/android/mediapicker/model/MediaPickerContext.kt).

**Merge Instructions**

* Update the version to `0.8.0`
* Merge the PR
* Create `0.8.0` release
* Manually publish it to S3 Maven
* Open PRs for each repository to update the plugin version to `0.8.0` and merge it before we merge any of the AGP update PRs

The reason I am suggesting we merge the `publish-to-s3-gradle-plugin` update PRs before AGP PRs is that, the AGP update PRs don't seem to have the warning anymore, so it becomes harder to verify these changes.

---

After all this work, I'll still note that this change was not strictly necessary for us to do at the moment. I think the warning looks a bit scarier than it actually is because even after AGP `8.0` update, the `release` component was still created as expected. We were also ahead of the curve on publishing the `-sources.jar` file using this plugin. However, there is no guarantee that a further AGP update wouldn't have broken our publishing flow. So, I think it is still a good idea to handle this before it becomes a problem for us.
